distributed/worker_state_machine.py

```diff
@@ -3591,16 +3574,99 @@ def __init__(self, state: WorkerState):
         self.state = state
         self._async_instructions = set()
 
-    def _handle_stimulus_from_task(self, task: asyncio.Task[StateMachineEvent]) -> None:
-        """An asynchronous instruction just completed; process the returned stimulus."""
+    def _start_async_instruction(
+        self,
+        task_name: str,
+        func: Callable[P, Awaitable[StateMachineEvent]],
+        /,
+        *args: P.args,
+        **kwargs: P.kwargs,
+    ) -> None:
+        """Execute an asynchronous instruction inside an asyncio task"""
+        ledger = DelayedMetricsLedger()
+
+        @wraps(func)
+        async def wrapper() -> StateMachineEvent:
+            with ledger.record():
+                return await func(*args, **kwargs)
+
+        task = asyncio.create_task(wrapper(), name=task_name)
+        self._async_instructions.add(task)
+        task.add_done_callback(partial(self._finish_async_instruction, ledger=ledger))
+
+    def _finish_async_instruction(
+        self,
+        task: asyncio.Task[StateMachineEvent],
+        ledger: DelayedMetricsLedger,
+    ) -> None:
+        """The asynchronous instruction just completed; process the returned
+        stimulus.
+        """
         self._async_instructions.remove(task)
         try:
             # This *should* never raise any other exceptions
             stim = task.result()
         except asyncio.CancelledError:
             # This should exclusively happen in Worker.close()
             return
-        self.handle_stimulus(stim)
+        except BaseException:  # pragma: nocover
+            logger.exception("async instruction handlers should never raise!")
+            raise
+
+        with ledger.record():
+            # Capture metric events in _transition_to_memory()
+            self.handle_stimulus(stim)
+
+        self._finalize_metrics(stim, ledger)
+
+    def _finalize_metrics(
+        self,
+        stim: StateMachineEvent,
+        ledger: DelayedMetricsLedger,
+    ) -> None:
+        activity: tuple[str, ...]
+        coarse_time: str | Literal[False]
```

Comment:
Note: This is a big switch statement. I personally would likely implement this as a method of the various events. attaching the ownership of this to the event moves the implementation slightly closer to the `TracedEvent` of https://github.com/dask/distributed/pull/7619

However, I don't have a strong preference and don't mind having this as a method here. I think there is a case for both, I just wanted to pass along this observation.

Reply:
I know, I thought about it but felt that putting it in methods would be
1. too dispersive - you'd need to go through all of those classes to see the fine differences between their implementation
2. it would require a back pointer to WorkerStateMachine in the parameters, which while technically sound is not a pattern we've used anywhere else
3. We would need to implement a dummy in `RetryBusyWorkerEvent`, just in virtue of it being the exit event of an async instruction - or alternatively we'd need to insert an extra level of parent classes `parent EventAtEndOfAsyncInstruction(StateMachineEvent)` just for this.